### PR TITLE
refactor(app-stores): Decouple sidebar, databases/collections, and collection stats from instance refresh and use single source for coll stats COMPASS-5266

### PR DESCRIPTION
**Pull request by gribnoysup**
## Description
This PR addresses a few architectural details of Compass state in preparation for making the loading flow of the data more on-demand rather than upfront. The general idea is more or less a follow-up to the initial models refactoring:

- instance-store keeps main reference to the instance and is responsible for reacting to events and fetching/refreshing data
- plugins don't fetch anything themselves, instead they subscribe to changes on the instance model and emit events that instance store will react to and start a fetch/update

This means that we can have plugins decoupled from the full instance refresh and at the same time have a single place where the source of the instance state is stored. NB: I didn't change all possible places in Compass to have this instance model as a single source of truth as this would more or less require us to refactor the whole app, instead only the parts that are important for the TTFQ were changed in this PR:

- compass-home
- sidebar plugin and all its children plugins like deployment-awareness
- databases-collections plugin
- collection-stats plugin (okay so this one was not strictly required, but it bothered me too much that we literally have this data on the previous screen usually and then always fetch it again)

This PR contains no UI changes as this refactoring is already pretty massive in scope. To be able to merge this I put the new "lazy" fetching logic and global overlay removal behind a feature flag. Setting `process.env.COMPASS_NO_GLOBAL_OVERLAY = 'true'` in runtime will disable global overlay and will not fetch collection list or collection info until database or a full namespace is selected. Here's a small demo of all three loading behaviors:

|GA|New w/ global overlay|New w/o global overlay|
|---|---|---|
|global_overlay_old|new_new_global_overlay|no_overlay_new|

Even with a feature flag there are some subtle differences in how the app behaves right how that I'll need to double-check if they are okay to keep first. If needed we can fully re-implement current global overlay behavior with the new store.

### Checklist
- [x] Fix all the tests
- [x] ~Double-check with Claudia temporary changes in the global overlay behavior~ showed it to Max and we both agreed that it's better to hide it for now when it's loading and the global overlay is active

## Motivation and Context
Currently when you start Compass application, we load a lot of data upfront: this includes some instance info, databases list, their stats, and every collection list for every database. Separately these commands are fast enough, but combined together they might take awhile, preventing user from interacting with Compass when it might be ready for interaction (e.g., you are blocked by the overlay and waiting for the `listCollection` on a group of particularly big databases when the one you really wanted to interact with is in another database that already loaded this information).

We want to split this loading flow into separate independent steps that fire when appropriate, but in the current state of things is that a lot of plugins depend on global `instance-refreshed` event that will contain all that information. To be able to move forward with the idea of more asynchronous data flow this PR removes all dependencies on the `instance-refreshed` event and instead subscribes plugins to the separate bits of information that they require from the instance model
- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [x] Misc

## Open Questions
As always, my goal here is not to make Ampersand _the way_ of handling state in Compass, it's just a useful tool we have right now (better than trying to achieve the same goal using our redux/reflux stores) that we can throw away later as soon as we know better what exactly we want state management in Compass to be. But if you feel like this is too much Ampersand, we should discuss if we want to continue with these changes

### CODE REVIEW

**Review by gribnoysup — COMMENTED** [review]
Left a few comments in places I thought would benefit from having some additional context provided, but obviously whatever catches your attention in the code here, please don't hesitate to leave a comment

File: packages/compass-home/src/components/home.tsx
Comment (mcasimir) [inline packages/compass-home/src/components/home.tsx:240]: is this just giving a better name or the semantic is actually different? Like this event is meant to be happen only once?

File: packages/compass-import-export/src/modules/import.js
Comment (mcasimir) [inline packages/compass-import-export/src/modules/import.js:323]: Nice one :)

File: packages/compass-home/src/components/home.tsx
Comment (mcasimir) [reply packages/compass-home/src/components/home.tsx:166]: I agree with you, wouldn't change the state of things too much at this stage. Better to agree first on a blueprint to how things would look like long-term and leave everything as it is in the meanwhile.

File: packages/instance-model/lib/model.js
Comment (mcasimir) [inline packages/instance-model/lib/model.js:152]: this function, and all the other fetch functions in the model, are both stateful and async at the same time which may cause funny behaviours. We probably could avoid multiple concurrent calls to modify state with some kind of guard

**Review by mcasimir — COMMENTED** [review]
👍, left some comment, there is something about `fetch` functions being called contemporarily that could be worth to look into before merging

File: packages/compass-home/src/components/home.tsx
Comment (mcasimir) [reply packages/compass-home/src/components/home.tsx:166]: I agree with you, wouldn't change the state of things too much at this stage. Better to agree first on a blueprint to how things would look like long-term and leave everything as it is in the meanwhile.

File: packages/instance-model/lib/model.js
Comment (gribnoysup) [reply packages/instance-model/lib/model.js:152]: Makes sense, yeah! For context this kind of stuff already happens in Compass in a lot of places so I wasn't sure if I want to expand the scope and introduce even more changes in the models, but now that most of this fetching logic is in one place, we might as well do it 👍

File: packages/compass-home/src/components/home.tsx
Comment (gribnoysup) [reply packages/compass-home/src/components/home.tsx:240]: This even is different from the previous one. The `instance-refreshed` is called every time we "refresh" the instance which currently includes 1) instance created for the first time when user connects and initial data is fetched 2) user does something that triggers an instance refresh: pressing a refresh button, adding/removing documents, running an aggregation with an $out step, and some other things I don't remember right away.

This new `instance-created` event is emitted only when user connects and we create the instance, at this point we have not even started the initial fetch/refresh call. This gives other stores in the app the opportunity to subscribe to the partial state tree changes (e.g., only in database or collection they are interested in) instead of just flushing their state every time something in the app caused a state refresh

File: packages/compass-collection-stats/src/stores/store.js
Comment (gribnoysup) [reply packages/compass-collection-stats/src/stores/store.js:62]: Yeah, I was just not changing it from the initial implementation, but the diff here is messed up here anyway, so might as well clean it up. Thanks for the suggestion!

File: packages/compass-collection-stats/src/stores/store.js
Comment (mcasimir) [reply packages/compass-collection-stats/src/stores/store.js:8]: I guess it would be enough to add a comment, I started to think weird stuff, like, if is a symbol then reflux won't add a watcher etc .. 🤣

if we want a simpler way probably just `store._appRegistry` and `store._instance` 🤷‍♂️, but i'm ok with the symbol

**Comment by mcasimir** [comment]
Should we follow up with some loading indication when a database is uncollapsed? feels a bit like nothing is going on right now..  except for that it looks already amazing.

Just looking at the .gif, in the context of doing the first query, is there a way to get the connection names even earlier?

Looking at the screenshot it seems like db stats are getting populated before the collection names, makes me think we are probably consuming the connection pool to do stuff that we could delay to after all connection names are loaded? Don't mean to block this one, i believe this could just be something to consider afterward!

**Comment by gribnoysup** [comment]
> Should we follow up in another PR/ticket with some loading indication for when a database is uncollapsed? feels a bit like nothing is going on right now.. except for that it looks already amazing.

Maybe I'm misunderstanding but just to clarify this particular behavior, the one you can see on the "New w/o global overlay" gif is behind a feature flag mostly so we can test it and play around with it, this PR changes the loading logic and decouples a bunch of things internally, but by default if we land this change in `main` it will preserve the current global overlay and the way it loads data so you will see no jumps. After this change is reviewed, approved, and merged (and I'm done fixing all the gazillion tests I broke 😆), there are two additional JIRA tickets that will be specifically about adding loading states for the sidenav and databases/collections view and as soon as those are done we can disable overlay and remove it from the source. Doing this UI change in the same PR would just be too much so having a "feature flag" was the best way I could find to limit the amount of changes I'm doing in one go, sorry if it's causing confusion, I hope this clarifies things!

(I do need to update the "New w/ global overlay" gif as after talking with Max we decided that showing loaded databases while we are blocking the screen with overlay might be a bit annoying for the users, so now it's 1 to 1 the same behavior as it was before: we don't show anything until overlay is gone)

File: packages/compass-collection-stats/src/stores/store.js
Comment (gribnoysup) [reply packages/compass-collection-stats/src/stores/store.js:8]: I'll keep it but will add a comment! Thanks for highlighting this

**Comment by gribnoysup** [comment]
> Looking at the screenshot it seems like db stats are getting populated before the collection names, makes me think we are probably consuming the connection pool to do stuff that we could delay to after all connection names are loaded? Don't mean to block this one, i believe this could just be something to consider afterward!

This is what I meant when I said that it would be great if we could have a priority-based pool for these requests 😄 In theory we probably want `dbStats` in advance because we use them to show the collections count right now or we can use it to show the exact amount of placeholders for the collection names when you expand (otherwise we don't know anything about the count and avoiding jump in UI is not really possible). So if we could express in code something like: fetch all the db stats, but if we want to list collections, prioritize it, that would be pretty cool I think (and probably can be built on top of driver if we want to put some time into it?). Totally up for rethinking and improving the initial fetching flow further, but this will most definitely require some additional UI changes that we didn't thought through enough for this particular project

File: packages/instance-model/lib/model.js
Comment (gribnoysup) [reply packages/instance-model/lib/model.js]: Done in 8ddc00d, but now I'm starting to think if I should actually move all these models in one place due to all the code duplication between the three, any thoughts?

File: packages/compass-collection-stats/src/stores/store.js
Comment (gribnoysup) [reply packages/compass-collection-stats/src/stores/store.js]: Done in 9b8a800

File: packages/compass-collection-stats/src/stores/store.js
Comment (gribnoysup) [reply packages/compass-collection-stats/src/stores/store.js]: Done in 9b8a800b37730171a847b2fac66d1d831d67ef8c

File: packages/collection-model/lib/model.js
Comment (lerouxb) [inline packages/collection-model/lib/model.js:262]: I guess this os more a general codebase-wide style concern than something related to this specifically, but shouldn't this only swallow specific errors? Right now any error of any kind will cause it to fallback.

File: packages/compass-collection-stats/src/stores/store.js
Comment (lerouxb) [inline packages/compass-collection-stats/src/stores/store.js:125]: Not sure if this comment is really about this PR because I see this all over compass, but does it make sense for App.InstanceStore to not exist? What really happens if it doesn't? Do we test for that? Could we write things in such a way that we don't have these kinds of tests all over? This kind of thing introduces so much complexity with extra untested code paths.

File: packages/compass-collection-stats/src/stores/store.js
Comment (lerouxb) [inline packages/compass-collection-stats/src/stores/store.js:143]: Does it make sense to `store.updateCollectionDetails({})`? How would we get into that kind of situation and do we support it?

File: packages/compass-sidebar/src/components/sidebar/sidebar.jsx
Comment (lerouxb) [inline packages/compass-sidebar/src/components/sidebar/sidebar.jsx:146]: I don't really understand why db wouldn't have _id?

File: packages/collection-model/lib/model.js
Comment (gribnoysup) [reply packages/collection-model/lib/model.js:262]: I think most of the possible server errors that can be returned here we want to ignore and try to fallback to the `listCollectionsNamesOnly` just in case, I honestly would struggle to find all the possible ones returned from this particular server command, I don't think that any documentation specifically about that exists tbh. Are you thinking more about a case where our code is broken to the point that this throws something or do you think we should track down possible server/driver errors and match against them before proceeding?

Comment (lerouxb) [reply packages/collection-model/lib/model.js:262]: Is there some kind of common pattern on server errors we could check to see if it exists? Like a `code` property perhaps.

My thinking is that even a `Uncaught TypeError: Cannot read property 'foo' of undefined` will land here.

File: packages/compass-collection-stats/src/stores/store.js
Comment (gribnoysup) [reply packages/compass-collection-stats/src/stores/store.js:143]: We probably can't in any flow we can predict, but FWIW I don't think that it's inherently bad Compass doesn't end any weird condition it can run into with a hard crash, we are connecting it to basically whatever users want and don't really control the backend our app communicates with, so making code continue to run in some cases where it's not that critical doesn't really bother me. I am happy and open to get another perspective on that though

File: packages/collection-model/lib/model.js
Comment (gribnoysup) [reply packages/collection-model/lib/model.js:262]: Yeah, this makes sense and we probably can rely on `code` being there for that check, I'll dig into driver doc to confirm

Comment (lerouxb) [reply packages/collection-model/lib/model.js:262]: I wouldn't be too surprised if there's nothing on database errors to differentiate them from other errors in which case I guess you might as well leave it. Alternatively we could annotate db errors in data service or something where we can at least tell as close to the db as possible that it is a db error.

But even then if we had a bug like we're sending a query to the db that's invalid it will also just fall back without us knowing.

It might just be a wider problem we can't fix in this PR, though :)

File: packages/compass-sidebar/src/components/sidebar/sidebar.jsx
Comment (lerouxb) [reply packages/compass-sidebar/src/components/sidebar/sidebar.jsx:146]: OK sorry I misread that, then.

Comment (gribnoysup) [reply packages/compass-sidebar/src/components/sidebar/sidebar.jsx:146]: The logic for this collapsing is not obvious at a glance at all, I'll try my best to have tests that can highlight the use-cases _before_ I open a PR for review next time 😅

File: packages/compass-collection-stats/src/stores/store.js
Comment (lerouxb) [reply packages/compass-collection-stats/src/stores/store.js:125]: I don't know if we can really change that kind of pattern in compass at this point. I think it is a side-effect of how everything seems to be created at once and then they all listen to events. If we had more of a react component type pattern where we only create things once all their dependencies are there (ie. it is just props/params passed to it) then we wouldn't have this sort of thing all over the place, resulting in much simpler code with fewer branches. 

Changing it bit by bit at this stage is probably not really viable as there are just countless cases of this pattern (and similar) all over. Ideally I think code flow would never have gotten anywhere near a place like this where things it depends on have any chance of not existing.

File: packages/compass-collection-stats/src/stores/store.js
Comment (lerouxb) [reply packages/compass-collection-stats/src/stores/store.js:143]: I guess I'm just very averse to code that never executes and should never execute. Especially if we don't test or otherwise support those cases. I worry that it will just hide the real problems and cause hard to debug problems elsewhere. Generally I prefer to have things throw so we can at least find out about it. But I don't have enough context to know if this is just another case of the event based architecture I pointed out above - ie maybe we _can_ hit these conditions because other code is already written to be this forgiving and just continue rather than either be architected in such a way that these conditions aren't possible or at least throwing if the impossible happens.

Defensive/forgiving coding tends to cause a cascade of more defensive/forgiving coding elsewhere.

But that's just my opinion - feel free to ignore.

File: packages/compass-app-stores/src/stores/instance-store.js
Comment (lerouxb) [reply packages/compass-app-stores/src/stores/instance-store.js:63]: > I just don't trust events flying around in compass to the point that I'm always double-checking if stuff exists 

That's exactly the problem at this point - none of us can trust anything and it ends up infecting all the code and becomes a self-perpetuating problem. Very quickly we're all complicit ;)

Comment (gribnoysup) [reply packages/compass-app-stores/src/stores/instance-store.js:63]: I know, I know, and thanks for highlighting these, we definitely can do better when we encounter these

File: packages/compass-collection-stats/src/stores/store.js
Comment (gribnoysup) [reply packages/compass-collection-stats/src/stores/store.js:143]: Well I asked for the opinion, so I'm definitely not just ignoring it! 😄 I generally agree and as we touched on in a few other places it's just hard to do right now due to how Compass works, I will definitely pay more attention to this now and re-evaluate if we can handle some cases better in that sense, thanks for sharing your thoughts on this!

Comment (lerouxb) [reply packages/compass-collection-stats/src/stores/store.js:143]: My hope is that if we ever change this code to typescript it will force us to at least deal with some of these.

**Review by lerouxb — APPROVED** [review]

**Pull request merged**
